Objects/moduleobject.c

```diff
@@ -819,68 +872,109 @@ _Py_module_getattro_impl(PyModuleObject *m, PyObject *name, int suppress)
         Py_DECREF(getattr);
         return result;
     }
+
+    if (suppress == 1) {
+        return NULL;
+    }
     if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__name__), &mod_name) < 0) {
         return NULL;
     }
-    if (mod_name && PyUnicode_Check(mod_name)) {
-        PyObject *spec;
-        if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+    if (!mod_name || !PyUnicode_Check(mod_name)) {
+        Py_XDECREF(mod_name);
+        PyErr_Format(PyExc_AttributeError,
+                    "module has no attribute '%U'", name);
+        return NULL;
+    }
+    PyObject *spec;
+    if (PyDict_GetItemRef(m->md_dict, &_Py_ID(__spec__), &spec) < 0) {
+        Py_DECREF(mod_name);
+        return NULL;
+    }
+    PyObject *origin = NULL;
+    if (spec) {
+        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
+        if (rc == -1) {
+            Py_DECREF(spec);
             Py_DECREF(mod_name);
             return NULL;
         }
-        if (suppress != 1) {
-            int rc = _PyModuleSpec_IsInitializing(spec);
+        if (rc == 1 && !PyUnicode_Check(origin)) {
+            Py_DECREF(origin);
+            origin = NULL;
+        }
+    }
+
+    int is_possibly_shadowing = _is_module_possibly_shadowing(origin);
+    if (is_possibly_shadowing == -1) {
+        goto done;
+    }
```

Comment:
There are a couple of tricky cases here to consider:

1. for some modules `spec.origin` might be set to `None` (e.g. namespace packages)
2. for any package we should probably check the module's `__path__` (AKA `spec.submodule_search_paths`)

As to the first case, the code here (at line 901) treats `None` as though `spec.origin` were not set at all.  I think that's the best we can do (when such a module isn't a package).

For the second case (including namespace packages which also hit the first case), it may be most correct to check `_is_module_possibly_shadowing()` for each entry in the module's `__path__`.  For namespace packages that's all we could do.

<details>
    <summary>(expand for code)</summary>

```C
    PyObject *origin = NULL;
    PyObject *path = NULL;
    if (spec) {
        // It may be a package.
        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(submodule_search_paths), &path);
        if (rc == -1) {
            goto done;
        }
        if (path != NULL) {
            if (!PySequence_Check(path)) {  // maybe for PyMapping_Check() too?
                Py_DECREF(path);
                path = NULL;
            }
            else {
                PyObject *_path = path;
                path = PySequence_List(_path);
                Py_DECREF(_path);
                if (path == NULL) {
                    goto done;
                }
            }
        }

        int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
        if (rc == -1) {
            Py_XDECREF(path);
            goto done;
        }
        // We could special-case (origin == Py_None) here.
        if (rc == 1 && !PyUnicode_Check(origin)) {
            Py_DECREF(origin);
            origin = NULL;
        }
    }
    
    int is_possibly_shadowing = -1;
    if (path != NULL) {
        Py_ssize_t size = PyList_GET_SIZE(path);
        for (Py_ssize_t i=0; i < size; i++) {
            PyObject *entry = PyList_GET_ITEM(path, i);
            if (entry == NULL) {
                is_possibly_shadowing = -1;
                break;
            }
            is_possibly_shadowing = _is_module_possibly_shadowing(entry);
            if (is_possibly_shadowing) {
                break;
            }
        }
        Py_DECREF(path);
        if (is_possibly_shadowing == -1) {
            goto done;
        }
        if (!is_possibly_shadowing) {
            is_possibly_shadowing = -1;
        }
    }
    if (is_possibly_shadowing == -1) {
        is_possibly_shadowing = _is_module_possibly_shadowing(origin);
        if (is_possibly_shadowing == -1) {
            goto done;
        }
    }
```

</details>

For *typical* non-namespace packages we can get the same effect with `os.path.dirname(os.path.dirname(origin))`, rather than the plain `os.path.dirname(origin)` we are doing now.

Reply:
Here's a fun fact (that I don't think affects us here): frozen modules have `spec.origin` set to "frozen", but `__file__` *may* be set to the original filename.

Reply:
Another tricky case: source-less modules (e.g. where only the .pyc file was distributed).  We'd need to check `spec.cached` (or `mod.__cached__`).

Reply:
Thank you for the detailed breakdown, I'll re-read your suggestion and incorporate handling for packages. I do have a somewhat naive question — when is it possible for a namespace package to shadow something?

If not, I wonder if we can get away with: `is_module_possibly_shadowing(os.path.dirname(origin) if hasattr(mod, "__path__") else origin)`

>  frozen modules have spec.origin set to "frozen"

Yeah, that is interesting. I agree it doesn't affect us here, and your other suggestion to use `has_location` means we're much more likely to be correct in situations like that

> source-less modules

I hadn't consider that, I'll play around with it

Reply:
> when is it possible for a namespace package to shadow something?

Given the following filesystem tree:

```
myproject/
    collections/
        spam.py
    numpy/
        eggs.py
    # subtle: a package can shadow a non-package module, and vice versa
    threading/
        ham.py
    myscript.py
```

and myscript.py:

```
import collections
import numpy as np
import threading

nt = collections.namedtuple('Record', 'a b c')
arr = np.array([1, 2, 3])
t = threading.Thread(target=lambda: None)
...
```

When running `python3 myproject/myscript.py`, all three namespace packages would shadow the corresponding stdlib/PyPI modules, wouldn't they?

The same directory structure would lead to the same shadowing if the "myproject" directory were manually placed at the beginning of `sys.path`.

Note that this isn't any different from other modules/packages.  The only difference with a namespace package is that it may be spread across multiple `sys.path` entries, rather than just one.

Reply:
No, they won't shadow the stdlib / PyPI modules:

Reply:
No, namespace packages won't shadow the stdlib / PyPI modules:
```
~/dev/cpython/tmp_repo stdlib-error λ tree
.
├── collections
│   └── spam.py
├── cwd_is_on_path.py
└── threading
    └── ham.py

3 directories, 3 files

~/dev/cpython/tmp_repo stdlib-error λ ../python.exe -c 'import cwd_is_on_path; import collections; print(collections.__spec__); import threading; print(threading.__spec__)' 
ModuleSpec(name='collections', loader=<_frozen_importlib_external.SourceFileLoader object at 0x10351fc10>, origin='/Users/shantanu/dev/cpython/Lib/collections/__init__.py', submodule_search_locations=['/Users/shantanu/dev/cpython/Lib/collections'])
ModuleSpec(name='threading', loader=<_frozen_importlib_external.SourceFileLoader object at 0x1035c5af0>, origin='/Users/shantanu/dev/cpython/Lib/threading.py')
```

This is because namespace packages are only created if we weren't able to find a normal package/module: https://peps.python.org/pep-0420/#specification

It's possible that module in a namespace package on sys.path[0] could shadow another module in a namespace package elsewhere, but that's a much more niche situation. I'll look into what it would take to detect that, although I might leave it for a follow-up. (I'll definitely add some handling for non-namespace packages though)

Reply:
Oh, right!  I'm glad you figured that out.  I should have remembered it. 😄

> It's possible that module in a namespace package on sys.path[0] could shadow another module in a namespace package elsewhere, but that's a much more niche situation. I'll look into what it would take to detect that, although I might leave it for a follow-up.

sounds good

Reply:
I believe the only thing left here is the handling of sourceless modules.  Thanks for taking care of the rest!

Reply:
It looks like sourceless modules just work, since they have an origin:
```
~/dev/cpython/srcless stdlib-error λ echo 'print("sourceless")' > random.py
~/dev/cpython/srcless stdlib-error λ ../python.exe -m compileall -b random.py                     
Compiling 'random.py'...
~/dev/cpython/srcless stdlib-error λ ll          
total 16
drwxr-xr-x    4 shantanu  staff   128B Mar  9 17:36 .
drwxr-xr-x  109 shantanu  staff   3.4K Mar  9 17:35 ..
-rw-r--r--    1 shantanu  staff    20B Mar  9 17:36 random.py
-rw-r--r--    1 shantanu  staff   146B Mar  9 17:36 random.pyc
~/dev/cpython/srcless stdlib-error λ rm random.py
~/dev/cpython/srcless stdlib-error λ ../python.exe -c 'import random; print(random.__spec__); print(random.randrange(10))'
sourceless
ModuleSpec(name='random', loader=<_frozen_importlib_external.SourcelessFileLoader object at 0x1015d83d0>, origin='/Users/shantanu/dev/cpython/srcless/random.pyc')
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    import random; print(random.__spec__); print(random.randrange(10))
                                                 ^^^^^^^^^^^^^^^^
AttributeError: module 'random' has no attribute 'randrange' (consider renaming '/Users/shantanu/dev/cpython/srcless/random.pyc' since it has the same name as the standard library module named 'random' and takes precedence over it on sys.path)
```

Does that sound right, or did you have another situation in mind? Sourceless isn't something I have much experience with